man/nsvc.8
.TH nsvc 8 "Jan 19, 2010"
.SH NAME
nsvc \- control ninit
.SH SYNOPSIS
.B nsvc
[ -Sservice ]
[
.B \-[uodgpchaitkorRDCLHVWZE]
] 
.I service
[\fI...\fR]
.br
.B nsvc  -Ppid
.I service
.
.SH DESCRIPTION
.B nsvc
is the management interface to ninit.
.I service
is the service directory name relative to /etc/ninit.  
You can also include /etc/ninit/ in the service name.
.PP
Each service directory contains control files.
They are described on 
.I http://riemann.fmi.uni-sofia.bg/ninit/

.PP
It is possible to make nsvc SUID.
.PP
.RS
.B chown root.root /bin/nsvc
.br
.B chmod 4711 /bin/nsvc
.RE
.PP
Then nsvc opens the pipes
.IR /etc/ninit/in , 
.I /etc/ninit/out 
and switches to real UID.
If UID is nonzero only limited set of operation are allowed.
.
.SH OPTIONS
If no options are given,
.B nsvc
will just print a diagnostic message to stdout, saying if the
service is up, down or finished, which PID it has if it is up, and for
how long it has been in this state.

Only the service name 
.B ALL 
means all services.

.TP
.B \-Sservice
Skip service.  Apply this option immediately after
.B nsvc.  
Don't insert space between S and service.
Examples:
.br
.B nsvc
\-Sngetty \-Ssshd \-Slogger \-d ALL
.br
.B nsvc
\-Sngetty \-W3 ALL || 
.B nsvc
\-Sngetty \-k ALL

.TP
.B \-u \-uNumber
Up.
If the service is not running, start it.
If the service stops, restart it.  If
.B Number
is nonzero and the service is down start it after
.B Number
seconds.
.TP
.B \-o \-oNumber
Once.
If the service is down, start it.
If the service stops, do not restart it.
If
.B Number
is nonzero and the service is down, start it after
.B Number
seconds; if it is up, restart it later.
.TP
.B \-d
Down.
If the service is running, send it a TERM signal and then a CONT signal.
After it stops, do not restart it.
.TP
.B \-r
Stop respawn.  
Set respawn flag to OFF.  This does not start/stop the service.
.TP
.B \-R
Start respawn.  
Set respawn flag to ON.  This does not start/stop the service.
.TP
.B \-p
Pause.
Send the service a STOP signal.
.TP
.B \-c
Continue.
Send the service a CONT signal.
.TP
.B \-h
Hangup.
Send the service a HUP signal.
.TP
.B \-a
Alarm.
Send the service an ALRM signal.
.TP
.B \-i
Interrupt.
Send the service an INT signal.
.TP
.B \-t
Terminate.
Send the service a TERM signal.
.TP
.B \-k
Terminate.
Send the service a KILL signal.
.TP
.B \-g 
Get.  Output just the PID.
.TP
.B \-Ppid \fIservice\fR
Set pid of service.
.TP
.B \-D \fIservice\fR
Print dependencies.
This will print all the names of all the services that were started
because this services depended on them.  Example:
.br
.B nsvc -D default
.TP
.B \-D
Print ninit memory usage statistics.
.TP
.B \-H
Print history.
This will print the names of some recently spawned processes.
This is useful if you see a process looping (initialization fails and
ninit is restarting it all the time).
.TP
.B \-L
Print all services loaded in memory.
.TP
.B \-V
Print version
.TP
.B \-Wnumber
Wait at most 
.B number
seconds service(s) to finish.  Example:
.br
.B nsvc \-d qmail
.br
.B nsvc \-W180 qmail || nsvc -k qmail
.TP
.B \-Cnumber \-C+number
Tell ninit that the service is in CRON mode.
If 
.B number 
is zero it disables CRON mode.  Let now=`date +%s`.
If 
.B number < now 
then ninit starts the service immediately,
otherwise ninit will start it latter.  
The flag
.B \-C+number
(if the
.B number 
is positive)
is equaivalent to the following:
.br
.B t=`expr now + number`
.br
.B nsvc -C$t ...

To stop foo,
which is in CRON mode do:
.br
.B nsvc -C0 foo
.br
.B nsvc -d foo
.br
.\" To see the next CRON timestamp of foo type:
.\" .B nsvc foo
.TP
.B \-Z \-Znumber
Zero (free) a service.  This is done with the ninit-reload program.
Example:
.br
.B nsvc -Z30 foo
.br
This removes foo (if such a service exists)
and prepares ninit to accept 30 new services approximately.
.TP
.B \-E \-Enumber
Update ninit environ.  This is done by ninit-reload program.
Example: 
.br
.B nsvc \-E36 ABC=12 UVW
.br
This updates the variable 
.B ABC 
and removes 
.B UVW.
It prepares ninit to accept 36 new services approximately.
See the environ with:
.br
.B tr '\\\\000' '\\\\012' < /proc/1/environ 

.
.SH "RETURN CODES"
Generally, 
.B nsvc 
returns zero if everything is OK or 1 on error (could not
open /etc/ninit/in or /etc/ninit/out or there is no process with the
given name).  In diagnostic mode, it will exit 0 if the service is up, 2
if it is down or 3 if it is finished.

.
.SH "ENVIRON"
.B nsvc
uses the variables 
.B NINIT_MEMORY
, 
.B NINIT_HOME
,
and 
.B NINIT_SYSDIR

.SH "SEE ALSO"
ninit(8), ninit-scan(8)
.br
ninit-shutdown(8), nkillall(8), svc(8), msvc(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


nsvc.c
#define SVC_PROGRAM
#include <fcntl.h>
#include <time.h>
#include <sys/file.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include "ninit.h"
#include "nsvc_buffer.h"

extern int ioctl();

static char **Names_skip, **NamesX, *buf, *initroot, *initsysdir, *nsvc_other, respmode;
static int NamesI, do_update, tty = -111;

static void e(int n) { buffer_flush(buffer_1); _exit(n); }
static void die_xx(char *s) { errmsg_iam("nsvc"); err(2,s,0); _exit(1); }

#include "get_services.h"
#include "findservice.h"
#include "open_inout.h"
#include "tryservice_nsvc.h"

static char *get_name() {
  char *x, *s, **skp;
 again:
  if (NamesI >= 0) {
    x = (NamesI <= maxprocess) ? root_name(NamesI) : 0;
    ++NamesI;
  } else {
    x = *NamesX;
    ++NamesX;
  }
  if (x)
    for (skp=Names_skip; (s=*skp); skp++)
      if (!str_diff(s,x)) goto again;
  return x;
}

static void close_inout() {
  close(infd); infd = -1;
  close(outfd); outfd = -1;
}

static int __findservice(char *service) {
  int idx;
  get_services();
  idx=findservice(service);
#ifdef CLEAN_SERVICE
  if (idx<0) idx=findservice(cleanservice(service));
#endif
  return idx;
}

void print_pid(char *service, pid_t pid) {
  errmsg_iam("nsvc");
  carp(service,(pid < 0) ? ": no such service" : ": service not running");
}

/* return PID, -1 if service is not loaded */
static pid_t __readpid(char *service, int verbose) {
  pid_t idx=__findservice(service);
  if (idx >= 0) idx = root[idx].pid;
  if (verbose && idx<2) 
    print_pid(service, idx);
  return idx;
}

static char *fu_x(uint32t u, int len) {  /* len must be max 10 !!! */
  char *x = fu(u);
  int y = str_len(x);
  while (y < len) { *--x = ' '; y++; }
  return x;
}

static void fs_x(const char *s, int len) {
  int k = str_len(s);
  outs(s);
  if (k >= len) outc('\t');
  else while (k++ < len) outc(' ');
}

static void fmt_time(unsigned long up) {
  unsigned long u;
  u = up % 60; up /= 60;
                                   fmt_color(fu_x(u, 4),'6'," ");
  if (up) { u = up % 60; up /= 60; fmt_color(fu_x(u, 2),'4'," "); }
  if (up) { u = up % 24; up /= 24; fmt_color(fu_x(u, 2),'5'," "); }
  if (up)                          fmt_color(fu_x(up,4),'2',"");
}

static int fmt_service(int i, int flagtty) {
  struct process *pr = root + i;
  pid_t pid = pr->pid;
  time_t cr, uptime;
  static time_t now;

  if (tty==-111) {
#if defined(INIT_SYSTEM)
    pid_t dummy;
    tty = (ioctl(1, X_TIOCGPGRP, &dummy) == 0);
#else
    tty=isatty(1);
#endif
  }

  if (!now) now=time(0);
  uptime = now - pr->startedat;
  if (uptime<0) uptime=0;

  if (flagtty || tty) {
    char *name=root_name(i);
    if (flagtty<=4 && tty) {
      char flags[8], *x, ch;
      int len;

      if (errmsg_argv0) { outs(errmsg_argv0); outs(": "); }
      fs_x(name, 16);

      switch (pid) {
        case 0: x="  down"; ch='7'; break;
        case 1: x="finish"; ch='2'; break;
        default:  
	  if (pr->pr_respawn) ch='1'; else ch='4';
	  x=fu_x(pid,6);
      }
      fmt_color(x,ch," ");

      len=0;
      if (pr->pr_respawn)	flags[len++] = 'R';
      if (pr->cron)	flags[len++] = 'C';
      if (pr->pr_end)	flags[len++] = 'E';
      flags[len] = 0 ;
      fs_x(flags, 3);

      fmt_time(uptime);

      if ((cr = pr->cron)) {
	fmt_color("\tnext cron",'1',":");
	if (cr > now) fmt_time(cr-now);
	else outs("  waititing");
	outc('\t');
	fmt_color(fu(cr), '8', "");
      }
      outc('\n');
    }
    else
      msg(name);
  } else {
    errmsg_iam(0);
    msg(fu(pid)," ",fu(uptime));
  }

  if (pid==0) return 2; else if (pid==1) return 3; else return 0;
}

static void dumphistory() {
  int j;
  if (!maxhistory) {
    carp("ninit compiled without history support");
    return;
  }
  errmsg_iam(0);
  for (j=0; j<maxhistory; j++)
    if (--history[j] <= maxprocess) {
      fmt_service(history[j], 4+root[history[j]].pr_circular);
      root[history[j]].pr_circular = 1;
    }
}

static void dumpdependencies(char* service, unsigned int skip) {
  int i,idx;
  idx=__findservice(service);
  if (idx<0) { print_pid(service, -1); return; }

  errmsg_iam(0);
  for (i=0; i<=maxprocess; ++i) {
    if (root[i].father==idx && i!=idx) {
      char *name = root_name(i);
      int k=skip;
      while (k-->0) outs("  ");
      msg(name);
      if (skip<16) dumpdependencies(name,skip+1);
      else msg(buf,"  ...",name);
    }
  }
}

static void wait_childs(int len) {
  unsigned wtime;
  pid_t i, pids[len];
  char *x;

  for (len=0; (x=get_name());)
    if ((pids[len]=__readpid(x, 0)) > 1) ++len;

  close_inout();
  wtime = 4*atoulong(nsvc_other);

  while (1) {
    for (i=len; i>0;)
      if (kill(pids[--i],0) && errno != EPERM)
	pids[i] = pids[--len];
    if (len==0) _exit(0);
    if (wtime--==0) _exit(1);
    nano_sleep(0,250000000);
  }
}

static int try_srv(char *s, char mode) {
  int r;
  char *save = nsvc_other;
  if (mode == 'r') 
    nsvc_other = (respmode) ? "c" : 0;	/* must be odd char !!! */ 
  r = tryservice(s, mode);
  nsvc_other = save;
  return r;
}

static void get_timestamp() {
  char *x = nsvc_other;
  unsigned int now;
  if (x[-1] == 'C') {
    if  (x[0] == '+') ++x;
    else return;
  }
  if ((now = atoulong(x)))
    nsvc_other = fu(time(0) + now);
}

int main(int argc,char *argv[]) {
  int i,idx,ret=0;
  pid_t pid;
  char *x, **skp;

  __b_1.x = alloca(BUFFER_1_LEN);
  buf = alloca(BUFFER_TMP_LEN);
  mem.a = INIT_ALLOC_BUFFER;

  x = env_get("NINIT_MEMORY");
  if (x) mem.a = atoulong(x);
  mem.r=mem.a;
  if (mem.a==0 || mem.a > 64000 || (mem.x=alloca(mem.a))==0) _exit(1);

  errmsg_iam("nsvc");
  infd = -1; outfd = -1;
  INIT_ENV_GET_HOME(initroot,"NINIT_HOME","INIT_HOME");
  INIT_ENV_GET_HOME(initsysdir,"NINIT_SYSDIR","INIT_SYSDIR");

  Names_skip = alloca(argc * sizeof(char *));
  for (skp=Names_skip; (x=argv[1]); argv++) {
    if (x[0] != '-' || x[1] != 'S') break;
    *skp++ = x+2;
  }
  *skp = 0;

  {
    int uid=getuid(), euid=geteuid();
    if (initroot && (uid != euid)) { ops: die_xx("\ago away!"); }
    if (initroot==0) initroot=INITROOT;
	if (initsysdir==0) initsysdir=INITSYSDIR;

    if (argv[1] == 0) { setuid(uid); goto REMOVE; }
    open_inout(initroot);
    
    if (uid != euid) {
      if (setuid(uid)) goto ops;
      get_services();
      close_inout();
    }
  }

  nsvc_other = argv[1] + 2;

  if (argv[2]==0) {
    char *a1 = argv[1];
    get_services();
    close_inout();

    if (a1[0] == '-') {
      switch (a1[1]) {
      case 'V': goto REMOVE;
      case 'L':
	errmsg_iam(0);
	while (get_name()) fmt_service(NamesI-1, 1);
	e(0);
      case 'H':
	dumphistory();
	e(0);
      case 'D':
	msg("memory usage: ",fu(mem.l + root_names_len)," = ",
	    fu(mem.l),"+",fu(root_names_len)," = services+names [",
	    fu(maxprocess+1),"]");
	e(0);
      }
    }

    i=__findservice(a1);
    if (i==-1) { print_pid(a1, -1); e(1); }
    ret=fmt_service(i,0);
    
    if (i >= 0 && tty && root[i].pid > 1) {
      INIT_ARGS2(argv, "./procfs", fu(root[i].pid));
      goto PROCFS;
    }
  } else {
    char *x;
    unsigned int sig=0;

    x = argv[2];

    if (argv[1][0]=='-') {
      char flagdown=1, try_flag='r';
      respmode = 0;

      switch (argv[1][1]) {
      case 'P': 
	ret=try_srv(x, 'P');
	if (ret) carp("Could not set PID of service ",x);
	break;

      case 'D':	dumpdependencies(x,0); break;
      case 'E':
      case 'Z':
      REMOVE:
	argv[0] = "./remove";	/* ~~XXX: make sys as environ!~~ */
      PROCFS:
	close_inout();
	buffer_flush(buffer_1);
	ret = chdir(initsysdir);
	if (!ret) ret=execve(argv[0], argv, environ);
	break;
      default:

	if (argv[3] == 0 && !str_diff(argv[2], "ALL")) {
	  get_services();
	  argc = maxprocess + 4;
	} else {
	  NamesX = argv + 2;
	  NamesI--;
	}

	switch (argv[1][1]) {
	case 'W':
	  wait_childs(argc);
	case 'g':
	  while ((x=get_name())) {
	    pid=__readpid(x,1);    
	    if (pid > 1) { errmsg_iam(0); msg(fu(pid)); }
	    else ret = 1;
	  }
	  break;
	case 'u': respmode=1;
	case 'o':
	  try_flag='s';
	  goto try_it;

	case 'C': try_flag='c'; /* cron */
	case 'R': respmode=1;

	try_it:
	case 'r': flagdown=0;
	case 'd':
	  get_timestamp();
	  while ((x=get_name())) {
	    idx=try_srv(x, try_flag);
	    if (idx==0) {
	      if (try_flag=='s') idx=try_srv(x,'r');
	      else if (flagdown) {
		pid=__readpid(x,0);
		if (pid>1 && 0==kill(pid,SIGTERM)) kill(pid,SIGCONT);
	      }
	    }
	    if (idx) { print_pid(x,-1); ret=1; }
	  }
	  break;
	  
	default: 	  
#ifdef NSVC_SIGNAL_CODED
	  {
	    unsigned char *S, *Sig = (unsigned char *)NSVC_SIGNAL_CODED;
	    for (S=Sig; *S; S += 2)
	      if (argv[1][1] == *S) { sig = S[1]; break; }
	  }
#else
#define kk(C,S) case C: sig=S; break
	  kk('t',SIGTERM); kk('a',SIGALRM); kk('p',SIGSTOP); kk('c',SIGCONT);
	  kk('h',SIGHUP);  kk('i',SIGINT);  kk('k',SIGKILL);
#endif
	} /* inner switch */
      }   /* outer switch */ 
    }

    if (sig) {
      while ((x=get_name())) {
	pid=__readpid(x,1);
	if (pid < 2) 
	  ret=1;
	else if (kill(pid,sig)) {
	  char* s;
	  switch (errno) {
	  case EINVAL: s="invalid signal"; break;
	  case EPERM: s="permission denied"; break;
	  case ESRCH: s="no such pid"; break;
	  default: s="unknown error";
	  }
	  carp(x,": could not send signal ",fu(sig)," to PID ",fu(pid),": ",s);
	  ret=1;
	}
      }
    }
  }
  e(ret);
  return 0;
}
